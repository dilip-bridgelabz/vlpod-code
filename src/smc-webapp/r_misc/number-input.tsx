/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

import * as React from "react";
const { Col, Row, Input } = require("antd");

interface Props {
  number: number;
  min?: number;
  max?: number;
  on_change: (n: number) => void;
  unit?: string;
  disabled?: boolean;
}

interface State {
  number: number | string;
}

export class NumberInput extends React.Component<Props, State> {
  constructor(props) {
    super(props);
    this.state = { number: this.props.number };
  }

  componentWillReceiveProps(next_props) {
    if (this.props.number !== next_props.number) {
      this.setState({ number: next_props.number });
    }
  }

  saveChange = (e?) => {
    if (e != undefined) {
      e.preventDefault();
    }
    let n = this.state.number;
    if (typeof n == "string") {
      n = parseInt(n);
      if (isNaN(n)) {
        n = this.props.number;
      }
    }
    if (this.props.min != null && n < this.props.min) {
      n = this.props.min;
    } else if (this.props.max != null && n > this.props.max) {
      n = this.props.max;
    }
    this.setState({ number: n });
    this.props.on_change(n);
  };

  render() {
    const unit = this.props.unit != undefined ? `${this.props.unit}` : "";
    return (
      <Row gutter={16}>
        <Col xs={16}>
          <Input
            type="text"
            ref="input"
            value={
              this.state.number != undefined
                ? this.state.number
                : this.props.number
            }
            onChange={(e) =>
              this.setState({
                number: e.target.value,
              })
            }
            onBlur={this.saveChange}
            onKeyDown={(e) => {
              if (e.keyCode === 27) {
                // async setState, since it depends on props.
                this.setState((_, props) => {
                  number: props.number;
                });
              } else if (e.keyCode === 13) {
                this.saveChange();
              }
            }}
            disabled={this.props.disabled}
          />
        </Col>
        <Col xs={8} className="lighten">
          {unit}
        </Col>
      </Row>
    );
  }
}
