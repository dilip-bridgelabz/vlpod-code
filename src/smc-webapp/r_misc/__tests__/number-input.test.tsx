/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

import * as React from "react";
import { shallow } from "enzyme";
import { NumberInput } from "../number-input";

describe("smoke testing", () => {
  test("renders with required values", () => {
    const mock_change = jest.fn();
    const starting_num = 0;
    const min = 10;
    const max = 20;
    const updated_number = 12;

    const rendered = shallow(
      <NumberInput
        number={starting_num}
        min={min}
        max={max}
        on_change={mock_change}
      />
    );
    expect(rendered).toMatchSnapshot();

    rendered.setProps({
      number: updated_number,
    });

    expect(rendered).toMatchSnapshot();
  });

  test("renders with a unit", () => {
    const mock_change = jest.fn();
    const starting_num = 0;
    const min = 10;
    const max = 20;
    const unit = "lbs";

    const rendered = shallow(
      <NumberInput
        number={starting_num}
        min={min}
        max={max}
        on_change={mock_change}
        unit={unit}
      />
    );
    expect(rendered).toMatchSnapshot();
  });

  test("renders disabled", () => {
    const mock_change = jest.fn();
    const starting_num = 0;
    const min = 10;
    const max = 20;
    const disabled = true;

    const rendered = shallow(
      <NumberInput
        number={starting_num}
        min={min}
        max={max}
        on_change={mock_change}
        disabled={disabled}
      />
    );
    expect(rendered).toMatchSnapshot();
  });
});

describe("input behavior", () => {
  test("should call on_change when blurred", () => {
    const mock_change = jest.fn();
    const starting_num = 12;
    const min = 10;
    const max = 20;

    const final_value = 15;

    const rendered = shallow(
      <NumberInput
        number={starting_num}
        min={min}
        max={max}
        on_change={mock_change}
      />
    );
    const form = rendered.find("Input").first();
    form.simulate("change", { target: { value: final_value } });
    form.simulate("blur");
    expect(mock_change.mock.calls.length).toBe(1);
    expect(mock_change.mock.calls[0][0]).toBe(final_value);
  });

  test("should cap at max", () => {
    const mock_change = jest.fn();
    const starting_num = 12;
    const min = 10;
    const max = 20;
    const over_max = 25;
    const final_value = max;

    const rendered = shallow(
      <NumberInput
        number={starting_num}
        min={min}
        max={max}
        on_change={mock_change}
      />
    );
    const form = rendered.find("Input").first();
    form.simulate("change", { target: { value: over_max } });
    form.simulate("blur");
    expect(mock_change.mock.calls.length).toBe(1);
    expect(mock_change.mock.calls[0][0]).toBe(final_value);

    // When passed as prop
    rendered.setProps({ number: over_max });
    form.simulate("blur");
    expect(mock_change.mock.calls.length).toBe(2);
    expect(mock_change.mock.calls[1][0]).toBe(final_value);
  });

  test("should floor at min", () => {
    const mock_change = jest.fn();
    const starting_num = 12;
    const min = 10;
    const max = 20;
    const under_min = 5;
    const final_value = min;

    const rendered = shallow(
      <NumberInput
        number={starting_num}
        min={min}
        max={max}
        on_change={mock_change}
      />
    );
    const form = rendered.find("Input").first();
    form.simulate("change", { target: { value: under_min } });
    form.simulate("blur");
    expect(mock_change.mock.calls.length).toBe(1);
    expect(mock_change.mock.calls[0][0]).toBe(final_value);

    // When passed as prop
    rendered.setProps({ number: under_min });
    form.simulate("blur");
    expect(mock_change.mock.calls.length).toBe(2);
    expect(mock_change.mock.calls[1][0]).toBe(final_value);
  });
});
