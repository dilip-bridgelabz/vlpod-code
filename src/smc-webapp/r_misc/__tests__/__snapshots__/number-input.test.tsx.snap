// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke testing renders disabled 1`] = `
<Row
  gutter={16}
>
  <Col
    xs={16}
  >
    <Input
      disabled={true}
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value={0}
    />
  </Col>
  <Col
    className="lighten"
    xs={8}
  />
</Row>
`;

exports[`smoke testing renders with a unit 1`] = `
<Row
  gutter={16}
>
  <Col
    xs={16}
  >
    <Input
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value={0}
    />
  </Col>
  <Col
    className="lighten"
    xs={8}
  >
    lbs
  </Col>
</Row>
`;

exports[`smoke testing renders with required values 1`] = `
<Row
  gutter={16}
>
  <Col
    xs={16}
  >
    <Input
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value={0}
    />
  </Col>
  <Col
    className="lighten"
    xs={8}
  />
</Row>
`;

exports[`smoke testing renders with required values 2`] = `
<Row
  gutter={16}
>
  <Col
    xs={16}
  >
    <Input
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value={12}
    />
  </Col>
  <Col
    className="lighten"
    xs={8}
  />
</Row>
`;
